File: app/src/main/java/com/pavelprymak/propodcast/presentation/screens/SearchFragment.java
package com.pavelprymak.propodcast.presentation.screens;


import android.os.Bundle;
import android.os.Handler;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.PopupMenu;
import androidx.appcompat.widget.SearchView;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProviders;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.pavelprymak.propodcast.App;
import com.pavelprymak.propodcast.MainActivity;
import com.pavelprymak.propodcast.R;
import com.pavelprymak.propodcast.databinding.FragmentSearchBinding;
import com.pavelprymak.propodcast.model.db.FavoritePodcastEntity;
import com.pavelprymak.propodcast.model.network.pojo.search.ResultsItem;
import com.pavelprymak.propodcast.presentation.adapters.SearchPodcastAdapter;
import com.pavelprymak.propodcast.presentation.adapters.SearchPodcastClickListener;
import com.pavelprymak.propodcast.presentation.viewModels.FavoritePodcastsViewModel;
import com.pavelprymak.propodcast.presentation.viewModels.SearchViewModel;
import com.pavelprymak.propodcast.utils.KeyboardUtil;
import com.pavelprymak.propodcast.utils.ShareUtil;

import java.util.ArrayList;
import java.util.List;

import static com.pavelprymak.propodcast.presentation.screens.PodcastDetailsFragment.ARG_PODCAST_ID;
import static com.pavelprymak.propodcast.utils.KeyboardUtil.showInputMethod;
import static com.pavelprymak.propodcast.utils.PodcastItemToFavoriteConverter.createFavorite;
import static com.pavelprymak.propodcast.utils.firebase.AnalyticsHelper.sentFirebaseAnalyticSearchQueryData;

public class SearchFragment extends Fragment implements SearchPodcastClickListener {
    private FragmentSearchBinding mBinding;
    private SearchViewModel mSearchViewModel;
    private SearchPodcastAdapter mAdapter;
    private FavoritePodcastsViewModel mFavoritePodcastsViewModel;
    private NavController mNavController;
    private List<FavoritePodcastEntity> mFavorites = new ArrayList<>();
    private Handler mDelayHandler = new Handler();
    private static final long SCROLL_DELAY = 300L;


    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        ((MainActivity) getActivity()).setNavViewVisibility(true);
        mFavoritePodcastsViewModel = ViewModelProviders.of(getActivity()).get(FavoritePodcastsViewModel.class);
        mSearchViewModel = ViewModelProviders.of(getActivity()).get(SearchViewModel.class);
        // Inflate the layout for this fragment
        mBinding = DataBindingUtil.inflate(inflater, R.layout.fragment_search, container, false);
        return mBinding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mNavController = Navigation.findNavController(view);
        prepareRecycler();
        mFavoritePodcastsViewModel.getFavorites().observe(this, favoritePodcastEntities -> {
            mFavorites.clear();
            if (favoritePodcastEntities != null) {
                mFavorites.addAll(favoritePodcastEntities);
            }
        });
        mSearchViewModel.getSearchResultsObserver().observe(this, resultsItems -> {
            mBinding.retryBtn.setVisibility(View.GONE);
            mAdapter = new SearchPodcastAdapter(this);
            mBinding.searchRecycler.setAdapter(mAdapter);
            mAdapter.submitList(resultsItems);
        });
        mSearchViewModel.getLoadData().observe(this, this::showProgressBar);
        mSearchViewModel.getErrorData().observe(this, throwable -> {
            if (throwable != null) {
                mBinding.retryBtn.setVisibility(View.VISIBLE);
            }
        });
        mSearchViewModel.getIsEmptyListData().observe(this, this::showEmptyList);

        mBinding.retryBtn.setOnClickListener(v -> {
            int size = mSearchViewModel.retryAfterErrorAndPrevLoadingListSize();
            mDelayHandler.postDelayed(() -> mBinding.searchRecycler.scrollToPosition(size - 1), SCROLL_DELAY);
        });

        mBinding.searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                String filterLanguage = App.mSettings.getFilterLanguage();
                mSearchViewModel.prepareSearchRequest(query, filterLanguage);
                KeyboardUtil.hideKeyboard(getActivity());
                sentFirebaseAnalyticSearchQueryData(query, filterLanguage);
                return true;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                return false;
            }
        });
        if (getResources().getBoolean(R.bool.isTablet)) {
            mBinding.fabFilter.hide();
        } else {
            mBinding.fabFilter.setOnClickListener(v -> {
                KeyboardUtil.hideKeyboard(getActivity());
                mNavController.navigate(R.id.languageFilterFragment);
            });
        }
        searchViewShowKeyboard();
    }


    private void searchViewShowKeyboard() {
        SearchView.SearchAutoComplete searchText = mBinding.searchView.findViewById(R.id.search_src_text);
        searchText.requestFocus();
        showInputMethod(getContext(), searchText);
    }

    private void showEmptyList(boolean isEmpty) {
        if (isEmpty) {
            mBinding.tvEmptySearch.setVisibility(View.VISIBLE);
        } else {
            mBinding.tvEmptySearch.setVisibility(View.GONE);
        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mFavoritePodcastsViewModel.getFavorites().removeObservers(this);
        mSearchViewModel.removeObservers(this);
    }

    private void prepareRecycler() {
        mAdapter = new SearchPodcastAdapter(this);
        LinearLayoutManager layoutManager = new LinearLayoutManager(getContext(), RecyclerView.VERTICAL, false);
        mBinding.searchRecycler.setLayoutManager(layoutManager);
        mBinding.searchRecycler.setAdapter(mAdapter);
    }

    private void showProgressBar(Boolean isShow) {
        if (isShow != null && isShow) {
            mBinding.progressBar.setVisibility(View.VISIBLE);
        } else {
            mBinding.progressBar.setVisibility(View.GONE);
        }
    }

    @Override
    public void onPodcastItemClick(String podcastId) {
        Bundle args = new Bundle();
        args.putString(ARG_PODCAST_ID, podcastId);
        mNavController.navigate(R.id.podcastDetailsFragment, args);
    }

    @Override
    public void onPodcastMoreOptionsClick(ResultsItem podcastItem, View v) {
        showPopupMenu(v, podcastItem);
    }

    private void showPopupMenu(View v, ResultsItem podcastItem) {
        if (getContext() == null) return;

        PopupMenu popupMenu = new PopupMenu(getContext(), v);
        popupMenu.inflate(R.menu.podcast_popup_menu);
        MenuItem menuItem = popupMenu.getMenu().findItem(R.id.action_favorite);
        boolean isFavorite = mFavoritePodcastsViewModel.isFavorite(mFavorites, podcastItem.getId());
        if (isFavorite) {
            menuItem.setTitle(R.string.remove_from_favorite);
        } else {
            menuItem.setTitle(R.string.add_to_favorite);
        }
        popupMenu
                .setOnMenuItemClickListener(item -> {
                    switch (item.getItemId()) {
                        case R.id.action_favorite:
                            if (isFavorite) {
                                mFavoritePodcastsViewModel.removeFromFavorite(podcastItem.getId());
                            } else {
                                mFavoritePodcastsViewModel.addToFavorite(createFavorite(podcastItem));
                            }
                            return true;
                        case R.id.action_share:
                            ShareUtil.shareData(getActivity(), podcastItem.getListennotesUrl());
                            return true;
                        default:
                            return false;
                    }
                });

        popupMenu.show();
    }
}

File: app/src/main/java/com/pavelprymak/propodcast/presentation/screens/SearchFragment.kt
package com.pavelprymak.propodcast.presentation.screens


import android.os.Bundle
import android.os.Handler
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.widget.PopupMenu
import androidx.appcompat.widget.SearchView
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.navigation.NavController
import androidx.navigation.Navigation
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.pavelprymak.propodcast.MainActivity
import com.pavelprymak.propodcast.R
import com.pavelprymak.propodcast.model.db.FavoritePodcastEntity
import com.pavelprymak.propodcast.model.network.pojo.search.ResultsItem
import com.pavelprymak.propodcast.presentation.adapters.SearchPodcastAdapter
import com.pavelprymak.propodcast.presentation.adapters.SearchPodcastClickListener
import com.pavelprymak.propodcast.presentation.viewModels.FavoritePodcastsViewModel
import com.pavelprymak.propodcast.presentation.viewModels.SearchViewModel
import com.pavelprymak.propodcast.utils.ApiErrorHandler
import com.pavelprymak.propodcast.utils.KeyboardUtil
import com.pavelprymak.propodcast.utils.KeyboardUtil.showInputMethod
import com.pavelprymak.propodcast.utils.PodcastItemToFavoriteConverter.createFavorite
import com.pavelprymak.propodcast.utils.SettingsPreferenceManager
import com.pavelprymak.propodcast.utils.ShareUtil
import com.pavelprymak.propodcast.utils.firebase.FirebaseAnalyticsHelper.sentAnalyticSearchQueryData
import kotlinx.android.synthetic.main.fragment_search.*
import org.koin.android.ext.android.inject
import org.koin.android.viewmodel.ext.android.sharedViewModel
import org.koin.android.viewmodel.ext.android.viewModel
import retrofit2.HttpException
import java.util.*


private const val SCROLL_DELAY = 300L
class SearchFragment : Fragment(), SearchPodcastClickListener {
    private val mSearchViewModel: SearchViewModel by viewModel()
    private val mFavoritePodcastsViewModel: FavoritePodcastsViewModel by sharedViewModel()
    private val mSettingsPref: SettingsPreferenceManager by inject()

    private lateinit var mNavController: NavController
    private var mAdapter: SearchPodcastAdapter? = null
    private val mFavorites = ArrayList<FavoritePodcastEntity>()
    private val mDelayHandler = Handler()


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        if (activity is MainActivity) {
            (activity as MainActivity).setNavViewVisibility(true)
        }
        return inflater.inflate(R.layout.fragment_search, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mNavController = Navigation.findNavController(view)
        prepareRecycler()
        mFavoritePodcastsViewModel.favorites.observe(this, Observer { favoritePodcastEntities ->
            mFavorites.clear()
            if (favoritePodcastEntities != null) {
                mFavorites.addAll(favoritePodcastEntities)
            }
        })
        mSearchViewModel.searchResultsObserver.observe(this, Observer { resultsItems ->
            retryBtn.visibility = View.GONE
            mAdapter = SearchPodcastAdapter(this)
            searchRecycler.adapter = mAdapter
            mAdapter?.submitList(resultsItems)
        })
        mSearchViewModel.loadData.observe(this, Observer<Boolean> { this.showProgressBar(it) })
        mSearchViewModel.errorData.observe(this, Observer { throwable ->
            throwable?.let {
                retryBtn.visibility = View.VISIBLE
                if (throwable is HttpException && context != null) {
                    ApiErrorHandler.handleError(context!!, throwable)
                }
            }
        })
        mSearchViewModel.isEmptyListData.observe(this, Observer<Boolean> { this.showEmptyList(it) })

        retryBtn.setOnClickListener { v ->
            val size = mSearchViewModel.retryAfterErrorAndPrevLoadingListSize()
            mDelayHandler.postDelayed({ searchRecycler.scrollToPosition(size - 1) }, SCROLL_DELAY)
        }

        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                val filterLanguage = mSettingsPref.filterLanguage
                mSearchViewModel.prepareSearchRequest(query, filterLanguage)
                KeyboardUtil.hideKeyboard(activity)
                sentAnalyticSearchQueryData(query, filterLanguage)
                return true
            }

            override fun onQueryTextChange(newText: String): Boolean {
                return false
            }
        })
        if (resources.getBoolean(R.bool.isTablet)) {
            fabFilter.hide()
        } else {
            fabFilter.setOnClickListener { v ->
                KeyboardUtil.hideKeyboard(activity)
                mNavController.navigate(R.id.actionFromSearchToLanguageFilter)
            }
        }
        searchViewShowKeyboard()
    }


    private fun searchViewShowKeyboard() {
        val searchText = searchView.findViewById<SearchView.SearchAutoComplete>(R.id.search_src_text)
        searchText.requestFocus()
        showInputMethod(context, searchText)
    }

    private fun showEmptyList(isEmpty: Boolean) {
        if (isEmpty) tvEmptySearch.visibility = View.VISIBLE else tvEmptySearch.visibility = View.GONE
    }

    override fun onDestroyView() {
        super.onDestroyView()
        mFavoritePodcastsViewModel.favorites.removeObservers(this)
        mSearchViewModel.removeObservers(this)
    }

    private fun prepareRecycler() {
        mAdapter = SearchPodcastAdapter(this)
        val layoutManager = LinearLayoutManager(context, RecyclerView.VERTICAL, false)
        searchRecycler.layoutManager = layoutManager
        searchRecycler.adapter = mAdapter
    }

    private fun showProgressBar(isShow: Boolean?) {
        if (isShow == true) progressBar.visibility = View.VISIBLE else progressBar.visibility = View.GONE
    }

    override fun onPodcastItemClick(podcastId: String) {
        val args = Bundle()
        args.putString(ARG_PODCAST_ID, podcastId)
        mNavController.navigate(R.id.actionFromSearchToDetails, args)
    }

    override fun onPodcastMoreOptionsClick(podcastItem: ResultsItem, v: View) {
        showPopupMenu(v, podcastItem)
    }

    private fun showPopupMenu(v: View, podcastItem: ResultsItem) {
        context?.let { context ->
            podcastItem.id?.let { podcastId ->
                val popupMenu = PopupMenu(context, v)
                popupMenu.inflate(R.menu.podcast_popup_menu)
                val menuItem = popupMenu.menu.findItem(R.id.action_favorite)
                val isFavorite = mFavoritePodcastsViewModel.isFavorite(mFavorites, podcastId)
                if (isFavorite) menuItem.setTitle(R.string.remove_from_favorite) else menuItem.setTitle(R.string.add_to_favorite)
                popupMenu
                    .setOnMenuItemClickListener { item ->
                        when (item.itemId) {
                            R.id.action_favorite -> {
                                if (isFavorite) mFavoritePodcastsViewModel.removeFromFavorite(podcastId) else mFavoritePodcastsViewModel.addToFavorite(
                                    createFavorite(podcastItem)
                                )
                                true
                            }
                            R.id.action_share -> {
                                ShareUtil.shareData(activity, podcastItem.listennotesUrl)
                                true
                            }
                            else -> false
                        }
                    }
                popupMenu.show()
            }

        }
    }
}
